### ISSUE TITLE
Example: How to create a multi-step job

### ISSUE THREAD

**Issue by kagkarlsson** [kagkarlsson/db-scheduler#245]
Add an example on how to create a multi-step job. 

Two possible solutions: 

* Maintain state in `task_data`, update and reschedule
* ~Schedule the next step at the very end of the current step using the SchedulerClient (not guaranteed atomic)~
* Schedule next step in the CompletionHandler

**Comment by heroInCommunity** [comment]
Hello @kagkarlsson ,

I am wondering how is the first option possible?
As I see, SchedulerClient.reschedule method is documented to throw an exception, if someone tries to invoke it from the current task.

Could you please give more details on the first option?

**Comment by frederikb** [comment]
Hi @kagkarlsson

Could you expand on why the second option (scheduling the next task at the end of the step) is not atomic? Doesn't saving the newly scheduled task happen in the same transaction as completing/deleting the current one?

**Comment by kagkarlsson** [comment]
In the second I refer to putting a `schedulerClient.schedule(...new execution for next step..)` at the end of the `ExecutionHandler`. The `ExecutionHandler` may run in a transactions, that is your code. But when it is done, the Scheduler runs, the `CompletionHandler` in its own transaction, for example removing the execution since it is done or something else. If that fails (very very unlikely, network problems), the next step will remain scheduled (separate transaction) but this current execution may run again, depending on behavior configured in the `DeadExecutionHandler`

**Comment by kagkarlsson** [comment]
Therefore I currently think the safest and simplest thing is putting the behavior in the completionhandler (i.e. when done, reschedule it with new data that indicates what step is next)

**Comment by kagkarlsson** [comment]
@heroInCommunity Sorry, I missed your questions.

> I am wondering how is the first option possible?
> As I see, SchedulerClient.reschedule method is documented to throw an exception, if someone tries to invoke it from the current task.

It is allowed (and expected) in the `CompletionHandler` (via `SchedulerOperations`). For example, see the code for the default `ComplationHandler` used by recurring tasks: 
https://github.com/kagkarlsson/db-scheduler/blob/master/db-scheduler/src/main/java/com/github/kagkarlsson/scheduler/task/CompletionHandler.java#L60

> Please correct me if I am wrong, in order to make steps running one after another and keep things transactional, we need to enable both options: enableImmediateExecution() and triggerCheckForDueExecutions().

It will not affect transactions, `enableImmediateExecution()` will remove the "pause" you would otherwise experience waiting for the next `pollingInterval`

**Comment by kagkarlsson** [comment]
I will write a quick example

**Comment by kagkarlsson** [comment]
Here is a quick example for using a CompletionHandler: https://github.com/kagkarlsson/db-scheduler/blob/master/examples/features/src/main/java/com/github/kagkarlsson/examples/JobChainingPocMain.java

Discovered that `enableImmediateExecution()` does not work for rescheduling in the CompletionHandler. Considering that a bug.

**Comment by kagkarlsson** [comment]
> Could you expand on why the second option (scheduling the next task at the end of the step) is not atomic? Doesn't saving the newly scheduled task happen in the same transaction as completing/deleting the current one?

The `CompletionHandler` is currently not transactional as it up until now only have had a single jdbc-statement. Might be needed to better support multistep-jobs..

**Comment by heroInCommunity** [comment]
Thank you @kagkarlsson , 
the first way looks promising, although we have implemented the second way at our project already.
We will consider refactoring to the first way in the future.

**Comment by kagkarlsson** [comment]
I may have been a bit quick to answer here:

> Could you expand on why the second option (scheduling the next task at the end of the step) is not atomic? Doesn't saving the newly scheduled task happen in the same transaction as completing/deleting the current one?

 Answer: 
> The CompletionHandler is currently not transactional as it up until now only have had a single jdbc-statement. Might be needed to better support multistep-jobs.

For the second option (in the description), the next step is actually scheduled at the very end of the `ExecutionHandler`. The `CompletionHAndler` will run after the full `ExecutionHandler` has run. If we put it at the very end of the `ExecutionHandler` there may be some edge-cases. The one I can think of is: 

* Execution A runs. Schedules execution B at the very end of its ExecutionHandler. Complete of execution A fail due to db/network-problems. Execution B remains scheduled and runs. Execution A is discovered as "dead" since it completion failed and the lock remained. Execution A is "revived" (depends on DeadExecutionHandler, customizable) and runs again after execution B have already run, and may actually schedule a new Execution B. To prevent you could for example make sure that a second Execution B will have no effect (make idempotent or similar)

**Comment by kagkarlsson** [comment]
After CompletionHandler is able to remove old and schedule new type of task in the same transaction, add another example, show that type of chaining.

**Comment by kagkarlsson** [comment]
Merged a PR enabling scheduling the next step in the CompletionHandler, safely. See [JobChainingUsingSeparateTasksMain.java](https://github.com/kagkarlsson/db-scheduler/blob/master/examples/features/src/main/java/com/github/kagkarlsson/examples/JobChainingUsingSeparateTasksMain.java)

I have also fixed so that immediateExecution works for scheduling that happens in the CompletionHandler, so `enableImmediateExecution()` will cause anything scheduled to `now()` (or before) to run directly, bypassing the `pollingInterval`.

I think what remains for these use-cases is reliably scheduling two or more new executions in a `CompletionHandler`. It is probably a niche use-case, but would be nice to support. Requires transactional CompletionHandlers.
